File: backend/src/main/java/com/example/demo/controller/CartController.java
package com.example.demo.controller;

import com.example.demo.dto.request.CartItemRequest;
import com.example.demo.dto.request.UpdateCartItemRequest;
import com.example.demo.dto.response.CartResponse;
import com.example.demo.response.ApiResponse;
import com.example.demo.service.interfaces.CartService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/**
 * Module 6 — Cart Management.
 *
 * The controller stays thin on purpose: validate the request, delegate to the
 * service, wrap the result. No business logic and no database access here.
 *
 * Every endpoint acts on the CURRENT user's cart — there is no cartId or userId
 * in any URL, which makes it impossible to touch someone else's cart.
 */
@RestController
@RequestMapping("/api/cart")
public class CartController {

    private final CartService cartService;

    public CartController(CartService cartService) {
        this.cartService = cartService;
    }

    /** GET /api/cart — current user's cart with totals. */
    @GetMapping
    public ResponseEntity<ApiResponse<CartResponse>> getCart() {
        CartResponse cart = cartService.getCart();
        return ResponseEntity.ok(ApiResponse.success("Cart retrieved successfully", cart));
    }

    /** POST /api/cart/items — add a product (201 Created). */
    @PostMapping("/items")
    public ResponseEntity<ApiResponse<CartResponse>> addItem(
            @Valid @RequestBody CartItemRequest request) {

        CartResponse cart = cartService.addItem(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Item added to cart", cart));
    }

    /** PUT /api/cart/items/{cartItemId} — change quantity / points on one line. */
    @PutMapping("/items/{cartItemId}")
    public ResponseEntity<ApiResponse<CartResponse>> updateItem(
            @PathVariable Integer cartItemId,
            @Valid @RequestBody UpdateCartItemRequest request) {

        CartResponse cart = cartService.updateItem(cartItemId, request);
        return ResponseEntity.ok(ApiResponse.success("Cart item updated", cart));
    }

    /** DELETE /api/cart/items/{cartItemId} — remove one line. */
    @DeleteMapping("/items/{cartItemId}")
    public ResponseEntity<ApiResponse<CartResponse>> removeItem(@PathVariable Integer cartItemId) {
        CartResponse cart = cartService.removeItem(cartItemId);
        return ResponseEntity.ok(ApiResponse.success("Cart item removed", cart));
    }

    /** DELETE /api/cart — empty the cart. */
    @DeleteMapping
    public ResponseEntity<ApiResponse<CartResponse>> clearCart() {
        CartResponse cart = cartService.clearCart();
        return ResponseEntity.ok(ApiResponse.success("Cart cleared", cart));
    }
}

File: dotnet-backend/Emart.Api/Controllers/CartController.cs
using System.Threading.Tasks;
using Emart.Api.DTOs;
using Emart.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Emart.Api.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    [Authorize]
    public class CartController : ControllerBase
    {
        private readonly ICartService _cartService;

        public CartController(ICartService cartService)
        {
            _cartService = cartService;
        }

        [HttpGet]
        public async Task<ActionResult<ApiResponse<CartDto>>> GetCart()
        {
            var cart = await _cartService.GetCartAsync();
            return Ok(ApiResponse<CartDto>.CreateSuccess("Cart retrieved successfully", cart));
        }

        [HttpPost("items")]
        public async Task<ActionResult<ApiResponse<CartDto>>> AddItem([FromBody] CartItemRequest request)
        {
            var cart = await _cartService.AddItemAsync(request);
            return StatusCode(201, ApiResponse<CartDto>.CreateSuccess("Item added to cart", cart));
        }

        [HttpPut("items/{cartItemId}")]
        public async Task<ActionResult<ApiResponse<CartDto>>> UpdateItem(int cartItemId, [FromBody] UpdateCartItemRequest request)
        {
            var cart = await _cartService.UpdateItemAsync(cartItemId, request);
            return Ok(ApiResponse<CartDto>.CreateSuccess("Cart item updated", cart));
        }

        [HttpDelete("items/{cartItemId}")]
        public async Task<ActionResult<ApiResponse<CartDto>>> RemoveItem(int cartItemId)
        {
            var cart = await _cartService.RemoveItemAsync(cartItemId);
            return Ok(ApiResponse<CartDto>.CreateSuccess("Cart item removed", cart));
        }

        [HttpDelete]
        public async Task<ActionResult<ApiResponse<CartDto>>> ClearCart()
        {
            var cart = await _cartService.ClearCartAsync();
            return Ok(ApiResponse<CartDto>.CreateSuccess("Cart cleared", cart));
        }
    }
}
